### ISSUE TITLE
Evaluation Protocol

### ISSUE THREAD

**Issue by fsnelgar** [Insta360-Research-Team/DiT360#24]
Thanks for your interesting work. 
There are some differences in performance reported here and former works such as PanFusion:

<img width="572" height="166" alt="Image" src="https://github.com/user-attachments/assets/3ea65228-4c4b-4778-a0a1-ad3254dfd8c8" />

<img width="421" height="74" alt="Image" src="https://github.com/user-attachments/assets/5efa5c48-55d5-4da2-9110-220e1aa7607a" />

In https://github.com/Insta360-Research-Team/DiT360/issues/23 you say you follow the same evaluation protocol as PanFusion, is it exactly the same?
* are the prompts the same (they're quite simple like 'the inside of a house') from blip3 here  [mp3d_stitched_caption.tar](https://monashuni-my.sharepoint.com/:u:/g/personal/cheng_zhang_monash_edu/Ec1A8tOmt_5ItvT2aktSUioBHzC_LRYjqaHPqipJuUhPHw?e=BgDGhL)
* do you generate 10 samples per prompt or just one? The original PanFusion pipeline is slightly unclear

If the evaluation code is available that would be very useful.

Thanks.

**Comment by fsnelgar** [comment]
Hi any update on this?
Also given the test size is ~1000 images, are the FID calculations (covariance on 2048 features) reliable?

**Comment by fenghora** [comment]
First of all, we sincerely apologize for only seeing your message and responding so late. Thank you very much for your interest, support, and helpful comments.

Our evaluation setup primarily follows **_PAR (Conditional Panoramic Image Generation via Masked Autoregressive Modeling)_**, a previous work by a former colleague, and we directly reused most of its evaluation settings. **_PanFusion_** was mainly used as an additional reference for the general evaluation procedure.

To the best of our recollection, the prompts were annotated using Qwen-VL. The corresponding prompts are included in the dataset that we released. For the reported evaluation, we generated one image for each prompt, rather than ten samples per prompt.

Below is a simplified version of our FID evaluation code containing the core implementation.

```
import os
from typing import Callable, List, Tuple

import torch
from PIL import Image
from torch.utils.data import DataLoader, Dataset
from torchmetrics.image.fid import FrechetInceptionDistance
from torchvision import transforms
from tqdm import tqdm


IMAGE_EXTENSIONS = (".png", ".jpg", ".jpeg")


def build_transform(
    output_size: Tuple[int, int],
) -> Callable[[Image.Image], torch.Tensor]:
    """
    Resize an RGB image and convert it to uint8 tensor in [0, 255].

    Args:
        output_size: Target size as (height, width).
    """
    return transforms.Compose(
        [
            transforms.Resize(
                output_size,
                interpolation=transforms.InterpolationMode.LANCZOS,
            ),
            transforms.ToTensor(),
            transforms.Lambda(
                lambda image: (image * 255).round().to(torch.uint8)
            ),
        ]
    )


class PairedImageDataset(Dataset):
    """
    Loads generated images and their corresponding reference images.

    Expected generated-image structure:
        generated_root/
            scene_id_view_id_.../
                generated_image.png

    Expected reference-image structure:
        reference_root/
            scene_id/
                matterport_stitched_images/
                    view_id_clean_big_dc.png
    """

    def __init__(
        self,
        generated_root: str,
        reference_root: str,
        output_size: Tuple[int, int],
    ) -> None:
        self.image_pairs = self._collect_pairs(
            generated_root=generated_root,
            reference_root=reference_root,
        )
        self.transform = build_transform(output_size)

        if not self.image_pairs:
            raise RuntimeError("No valid generated/reference image pairs were found.")

    @staticmethod
    def _find_first_image(folder: str) -> str | None:
        for filename in sorted(os.listdir(folder)):
            if filename.lower().endswith(IMAGE_EXTENSIONS):
                return os.path.join(folder, filename)
        return None

    def _collect_pairs(
        self,
        generated_root: str,
        reference_root: str,
    ) -> List[Tuple[str, str]]:
        pairs = []

        for subfolder in sorted(os.listdir(generated_root)):
            generated_folder = os.path.join(generated_root, subfolder)

            if not os.path.isdir(generated_folder):
                continue

            identifiers = subfolder.split("_")
            if len(identifiers) < 2:
                continue

            scene_id, view_id = identifiers[:2]
            generated_path = self._find_first_image(generated_folder)

            if generated_path is None:
                continue

            reference_path = os.path.join(
                reference_root,
                scene_id,
                "matterport_stitched_images",
                f"{view_id}_clean_big_dc.png",
            )

            if os.path.isfile(reference_path):
                pairs.append((generated_path, reference_path))

        return pairs

    def __len__(self) -> int:
        return len(self.image_pairs)

    def __getitem__(self, index: int) -> Tuple[torch.Tensor, torch.Tensor]:
        generated_path, reference_path = self.image_pairs[index]

        with Image.open(generated_path) as image:
            generated_image = self.transform(image.convert("RGB"))

        with Image.open(reference_path) as image:
            reference_image = self.transform(image.convert("RGB"))

        return generated_image, reference_image


@torch.inference_mode()
def compute_fid(
    generated_root: str,
    reference_root: str,
    output_size: Tuple[int, int] = (512, 1024),
    device: str = "cuda",
    batch_size: int = 32,
    num_workers: int = 8,
) -> float:
    """
    Compute FID between generated and reference images.
    """
    dataset = PairedImageDataset(
        generated_root=generated_root,
        reference_root=reference_root,
        output_size=output_size,
    )

    dataloader = DataLoader(
        dataset,
        batch_size=batch_size,
        shuffle=False,
        num_workers=num_workers,
        pin_memory=device.startswith("cuda"),
    )

    fid = FrechetInceptionDistance(
        feature=2048,
        normalize=False,
    ).to(device)

    for generated_images, reference_images in tqdm(
        dataloader,
        desc="Calculating FID",
    ):
        generated_images = generated_images.to(device, non_blocking=True)
        reference_images = reference_images.to(device, non_blocking=True)

        fid.update(reference_images, real=True)
        fid.update(generated_images, real=False)

    return float(fid.compute().item())


if __name__ == "__main__":
    score = compute_fid(
        generated_root="path/to/generated_images",
        reference_root="path/to/matterport3d",
        output_size=(512, 1024),
        device="cuda:0",
    )
    print(f"FID: {score:.4f}")
```

Regarding the reliability of FID with approximately 1,000 test images, we agree that this is a valid concern. Since the Inception features have 2,048 dimensions, estimating the feature covariance matrices from only around 1,000 samples can be statistically unstable. Therefore, the reported FID results should be treated mainly as a reference for comparison under the same evaluation setting, rather than as a highly precise absolute measurement.

Thank you again for pointing this out and for your constructive guidance.
